File: backend/api/tasks.py
from flask import jsonify
import mysql.connector
import os

# Inserts a new record into the "tasks" table
def insert_task(db, request):
    
    try:
        conn = db
        data = request.get_json()

        # Ensure required fields are present - if not, return a HTTP 400 error code
        if not all(key in data for key in ['title', 'description', 'author_id', 'assignee_id', 'due_date', 'complete']):
            return jsonify({"status": "error", "message": "Missing required fields"}), 400

        title = data['title']
        description = data['description']
        author_id = data['author_id']
        assignee_id = data['assignee_id']
        due_date = data['due_date']
        complete = data.get('complete', 0)

        cursor = conn.cursor()

        cursor.execute("""
            INSERT INTO tasks (title, description, author_id, assignee_id, due_date, complete)
            VALUES (%s, %s, %s, %s, %s, %s);
        """, (title, description, author_id, assignee_id, due_date, complete))

        # Fetch the result
        result = cursor.fetchall()

        conn.commit()
        cursor.close()
        conn.close()
        
        # Return the results in a structured response
        return jsonify({"status": "success", "data": result}), 200

    except mysql.connector.Error as e:
        # Handle database-specific errors
        print(f"Database error: {e}")
        return jsonify({"status": "error", "message": "Database error occurred"}), 500


    except Exception as e:
        # Handle general errors
        print(f"Error occurred: {e}")
        return jsonify({"status": "error", "message": "An unexpected error occurred"}), 500
    


# Get tasks assigned to a specific user (assignee_id included in GET request)
def get_user_tasks(db, request):

    try:
        # Retrieve assignee_id from URL query parameters
        assignee_id = request.args.get('assignee_id')

        # Ensure assignee_id is present
        if not assignee_id:
            return jsonify({"status": "error", "message": "Missing required fields: assignee_id"}), 400

        conn = db
        cursor = conn.cursor()

        cursor.execute("""
            SELECT * FROM tasks WHERE assignee_id = %s
        """, (assignee_id,))

        # Fetch the result
        result = cursor.fetchall()

        #conn.commit()
        cursor.close()
        conn.close()

        # Return the results in a structured response
        return jsonify({"status": "success", "data": result}), 200

    except mysql.connector.Error as e:
        # Handle database-specific errors
        print(f"Database error: {e}")
        return jsonify({"status": "error", "message": "Database error occurred"}), 500

    except Exception as e:
        # Handle general errors
        print(f"Error occurred: {e}")
        return jsonify({"status": "error", "message": "An unexpected error occurred"}), 500

File: brown-bottle-app/utils/api/tasks.ts
// This file will send a request to the backend python api to insert and get task info 

import Constants from 'expo-constants';

// Inserts a task into the database
export async function insertTask(title: string, description: string, author_id: number, assignee_id: number, due_date: string) {
  
  // Retrieve Environment Variables
  const { API_BASE_URL } = Constants.expoConfig?.extra || {};

  const base_url = API_BASE_URL;

  try {
    // Data you want to send in the request body
    const taskData = {
      title,
      description,
      author_id,   
      assignee_id,      
      due_date,
      complete: 0 // Default to incomplete/false (0)
    };

    const response = await fetch(`${base_url}/tasks/insert-task`, {
      method: "POST", // Using POST to send data
      headers: {
        "Content-Type": "application/json", // Ensure the backend understands the content type
      },
      body: JSON.stringify(taskData), // Send the task data in the request body as JSON
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    const data = await response.json();
    console.log(data); // Log the response data (including task_id)
    return data; // This will be the JSON response from Flask, containing the task ID

  } catch (error) {
    console.error("Failed to insert task:", error);
    throw error; // Optional: Rethrow the error if you want to handle it elsewhere
  }

}
  

export async function getUserTasks(assignee_id: string) {

  // Retrieve Environment Variables
  const { API_BASE_URL } = Constants.expoConfig?.extra || {};

  const base_url = API_BASE_URL;

  try {
    // Sending the assignee_id as a query parameter in the URL
    const response = await fetch(`${base_url}/tasks/get-user-tasks?assignee_id=${assignee_id}`, {
      method: "GET", // GET is appropriate for fetching data
      headers: {
        "Content-Type": "application/json", // Ensure the backend understands the content type
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    const data = await response.json();
    
    return data; // This will be the JSON response from Flask, containing the task ID

  } catch (error) {
    console.error("Failed to fetch user tasks:", error);
    throw error;
  }

}
